File: PAT_Basic_Level_by_lua/1031 查验身份证/1031 查验身份证.lua
-- 题目编号: 1031 查验身份证
-- 实现原理: 对前17位数字加权求和，结果模11得到校验码，与第18位比对。记录所有无效身份证号，全部通过则输出"All passed"。

-- 前17位数字对应的权重系数
local weights = { 7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2 }

-- 模11结果对应的校验码（索引从1开始，对应模11结果0~10）
local codes = { "1", "0", "X", "9", "8", "7", "6", "5", "4", "3", "2" }

-- 读取身份证数量 n，invalid 存储所有无效身份证号
local n, invalid = tonumber(io.read()), {}

-- 逐个检验每个身份证号
for _ = 1, n do
    local id, sum, valid = io.read("l"), 0, true
    
    -- 对前17位数字加权求和，并检查是否都是数字
    for i = 1, 17 do
        local d = tonumber(id:sub(i, i))
        -- 如果某一位不是数字，标记为无效
        if not d then
            valid = false
            break
        end
        sum = sum + d * weights[i]
    end
    
    -- 检查身份证是否有效：前17位都是数字 且 校验码正确
    if not valid or id:sub(18, 18) ~= codes[sum % 11 + 1] then
        invalid[#invalid + 1] = id
    end
end

-- 输出结果：全部通过则输出 "All passed"，否则逐行输出无效身份证号
if #invalid == 0 then
    print("All passed")
else
    for _, id in ipairs(invalid) do
        print(id)
    end
end

File: PAT_Basic_Level_by_ruby/1031 查验身份证/1031 查验身份证.rb
# 1031 查验身份证
#
# 题目描述：
#   一个合法的身份证号码由17位地区、日期编号和顺序编号加1位校验码组成。校验码的计算规则如下：
#   首先对前17位数字加权求和，权重分配为：{7，9，10，5，8，4，2，1，6，3，7，9，10，5，8，4，2}；然后将计算的和对11取模得到值`Z`；最后按照以下关系对应`Z`值与校验码`M`的值：
#   ```
#   Z：0 1 2 3 4 5 6 7 8 9 10
#   M：1 0 X 9 8 7 6 5 4 3 2
#   ```
#   现在给定一些身份证号码，请你验证校验码的有效性，并输出有问题的号码。
#
# 实现原理：
#   验证身份证号码的校验位是否正确；使用加权求和和模运算计算校验码；检查前17位是否全为数字
#

WEIGHTS = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2].freeze
CHECK_CODE = '10X98765432'.freeze

if __FILE__ == $PROGRAM_NAME
  n = gets.to_i
  invalid = []

  n.times do
    id = gets(chomp: true)
    sum = 0
    valid = true
    (0...17).each do |i|
      if id[i] < '0' || id[i] > '9'
        valid = false
        break
      end
      sum += id[i].to_i * WEIGHTS[i]
    end
    valid = false if valid && CHECK_CODE[sum % 11] != id[17]
    invalid << id unless valid
  end

  if invalid.empty?
    puts 'All passed'
  else
    puts invalid
  end
end
